[lldb-dap] Enabling instruction breakpoint support to lldb-dap.

santhoshe447 opened the pull request:
Added support for "supportsInstructionBreakpoints" capability and now it this command is triggered when we set instruction breakpoint.
We need this support as part of enabling disassembly view debugging. Following features should work as part of this feature enablement:

1. Settings breakpoints in disassembly view: Unsetting the breakpoint is not happening from the disassembly view. Currently we need to unset breakpoint manually from the breakpoint List. Multiple breakpoints are getting set for the same $

2. Step over, step into, continue in the disassembly view

The format for DisassembleRequest and DisassembleResponse at https://raw.githubusercontent.com/microsoft/vscode/master/src/vs/workbench/contrib/debug/common/debugProtocol.d.ts .

Ref Images:
Set instruction breakpoint in disassembly view:
![image](https://github.com/user-attachments/assets/833bfb34-86f4-40e2-8c20-14b638a612a2)

After issuing continue:
![image](https://github.com/user-attachments/assets/884572a3-915e-422b-b8dd-d132e5c00de6)

Commit:
[lldb][test] Add the ability to extract the variable value out of the summary.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command.

Adding a "port" property to the VsCode "attach" command likely extends the functionality of the debugger configuratiuon to allow attaching to a process using PID or PORT number.
Currently, the "Attach" configuration lets the user specify a pid. We tell the user to use the attachCommands property to run "gdb-remote <port>".
Followed the below conditions for "attach" command with "port" and "pid"
We should add a "port" property. If port is specified and pid is not, use that port to attach. If both port and pid are specified, return an error saying that the user can't specify both pid and port.

Ex - launch.json
{
	"version": "0.2.0",
    "configurations": [
        {
            "name": "lldb-dap Debug",
            "type": "lldb-dap",
            "request": "attach",
            "port":1234,
            "program": "${workspaceFolder}/a.out",
            "args": [],
            "stopOnEntry": false,
            "cwd": "${workspaceFolder}",
            "env": [],

        }
    ]
}

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command.

Adding a "port" property to the VsCode "attach" command likely extends the functionality of the debugger configuration to allow attaching to a process using PID or PORT number.
Currently, the "Attach" configuration lets the user specify a pid. We tell the user to use the attachCommands property to run "gdb-remote ".
Followed the below conditions for "attach" command with "port" and "pid"
We should add a "port" property. If port is specified and pid is not, use that port to attach. If both port and pid are specified, return an error saying that the user can't specify both pid and port.

Ex - launch.json
{
"version": "0.2.0",
"configurations": [
{
"name": "lldb-dap Debug",
"type": "lldb-dap",
"request": "attach",
"port":1234,
"program": "${workspaceFolder}/a.out",
"args": [],
"stopOnEntry": false,
"cwd": "${workspaceFolder}",
"env": [],

    }
]
}

In this patch we have resolved code formatting issues and fixed "test_by_name" failure.

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command.
Resolved code format issues.

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
Resolved code format issue.

Commit:
Resolved code format errors.

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command - Addressed review comments
Addressed all review comments.

Commit:
[lldb-dap] Added "port" property to vscode "attach" command - fix code format
resolved code format issues.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command - resolved all review comments.
All review comments have been addressed, and the attach-by-port tests have been verified on
Linux machine. Although the functionality is intended to support both Linux and macOS, we were
unable to verify it on macOS due to lack of access.

Commit:
[lldb-dap] Added "port" property to vscode "attach" command -resolved format check issues.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command. #91570
Resolved few review commennts.

Commit:
[lldb-dap] Added "port" property to vscode "attach" command.
Resolved code format issues.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command - addressed review comments.
Used named pipe to read debug server listening port.
Since we are only using "gdb-remote" process plugin, changed to "gdb-remote-port" and
"gdb-remote-hostname" from "port" and "hostname"

Commit:
[lldb-dap] Added "port" property to vscode "attach" command - resolved code format issues.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command - fixed review comments.

Commit:
[lldb-dap] Added "port" property to vscode "attach" command - resolved review comments.

Commit:
[lldb-dap] Added "port" property to vscode "attach" command.
I have shifted "Pipe" class to common location "lldb/packages/Python/lldbsuite/test/tools/lldb-server/lldbgdbserverutils.py"

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Updated README.md for newly added attach properties.

Added "gdb-remote-port" and "gdb-remote-hostname" attach properties usage in
README.md.

This was missed in PR https://github.com/llvm/llvm-project/pull/91570

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Updated README.md for newly added attach properties. #99926
Corrected spelling mistakes.

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Bump the version to 0.2.3

Bump the lldb-dap version to 0.2.3.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Added support for "supportsInstructionBreakpoints" capability and now it this command is triggered when we set instruction breakpoint.
We need this support as part of enabling disassembly view debugging. Following features should work as part of this feature enablement:

1. Settings breakpoints in disassembly view: Unsetting the breakpoint is not happening from the disassembly view. Currently we need to unset breakpoint manually from the breakpoint List. Multiple breakpoints are getting set for the same $

2. Step over, step into, continue in the disassembly view

The format for DisassembleRequest and DisassembleResponse at https://raw.githubusercontent.com/microsoft/vscode/master/src/vs/workbench/contrib/debug/common/debugProtocol.d.ts .

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Resolved code format issues.

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Resolved code format issues.

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
The following changes were committed by mistake, so they have been reverted.
"startDebugging", &g_dap.start_debugging_request_handler,
"repl-mode", &g_dap.repl_mode_request_handler,

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Few changes were missed in lldb/tools/lldb-dap/DAP.h, so added in this version.

walter-erquinigo commented on lldb/test/API/tools/lldb-dap/instruction-breakpoint/Makefile:
```diff
@@ -0,0 +1,6 @@
+CXX_SOURCES := main-copy.cpp
+CXXFLAGS_EXTRAS := -O1 -g
```
why don't you use -O0?

santhoshe447 replied:
I just verified with -O1, so same has been submitted.
I have changed it to "-O0"

walter-erquinigo commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -4046,6 +4048,181 @@ void request__testGetTargetBreakpoints(const llvm::json::Object &request) {
+  auto arguments = request.getObject("arguments");
+  auto breakpoints = arguments->getArray("breakpoints");
+
+  // It holds active instruction brealpoint list received from DAP.
```
```suggestion
  // It holds active instruction breakpoint list received from DAP.
```

walter-erquinigo reviewed (commented):
This LGTM, but I'd like @clayborg to give a second look.
Thanks for this fancy feature, I've been wanting it for a while

vogelsgesang commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -4078,6 +4255,9 @@ void RegisterRequestCallbacks() {
   g_dap.RegisterRequestCallback("threads", request_threads);
   g_dap.RegisterRequestCallback("variables", request_variables);
   g_dap.RegisterRequestCallback("disassemble", request_disassemble);
+  // Instruction breapoint request
```
we also need to update the stopped event to use `reason: instruction breakpoint` instead of `reason: breakpoint` in case we hit one of those breakpoints

santhoshe447 replied:
@vogelsgesang 
Thanks for bringing this point.
The GDB remote protocol does not have a specific packet for instruction breakpoints, which is likely why LLDB manages instruction breakpoints using the z0 packet, just like it does for software breakpoints.
The challenge is distinguishing between instruction breakpoints and software breakpoints when they use the same packet type.

**Instruction breakpoints packet info:**
$[{"name":"mandelbrot","reason":"breakpoint","registers":{"16":"3046555555550000","6":"10e3ffffff7f0000","7":"00e3ffffff7f0000"}],"signal":5,"tid":12342}]]#72

**Source breakpoint packet info:**
$[{"name":"mandelbrot","reason":"breakpoint","registers":{"16":"2e46555555550000","6":"10e3ffffff7f0000","7":"00e3ffffff7f0000"}],"signal":5,"tid":12342}]]#a6

It would be helpful if you could share any insights on this.

vogelsgesang replied:
I don't really know about those differences on the gdb protocol. I was only reading https://microsoft.github.io/debug-adapter-protocol/specification#Requests_SetInstructionBreakpoints

> When an instruction breakpoint is hit, a stopped event (with reason `instruction breakpoint`) is generated.

walter-erquinigo replied:
you could make something synthetic, and that could be very nice.
In the process event loop in lldb-dap, you could check if you stopped because of a breakpoint, and then you check if the breakpoint is an instruction breakpoint, and then you return that custom stop reason.
lldb-dap already does something like that in `CreateThreadStopped` for exception breakpoints.

santhoshe447 replied:
Thanks @walter-erquinigo 
This will work

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Resolved review comments.

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Added stop reason for instruction breakpoint.

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Added few comments

santhoshe447 committed:
Merge branch 'main' into main

ZequanWu commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -4082,6 +4084,260 @@ void request__testGetTargetBreakpoints(const llvm::json::Object &request) {
+      // Means delete removed breakpoint instance.
+      if (inst_reference == request_ibp.end()) {
+        g_dap.target.BreakpointDelete(prev_ibp.second.id);
+        removed_ibp.push_back(prev_ibp.first);
```
You can just remove it directly from instruction_breakpoints, right? There's no need to create the `removed_ibp` vector. Like: `g_dap.instruction_breakpoints.erase(name);`

santhoshe447 replied:
Thanks, removed_ibp is not required. I have updated the logic.

ZequanWu commented on lldb/tools/lldb-dap/JSONUtils.cpp:
```diff
@@ -766,6 +766,102 @@ llvm::json::Value CreateStackFrame(lldb::SBFrame &frame) {
+//   },
+//   "required": [ "id", "verified", "line"]
+// }
+llvm::json::Value CreateInstructionBreakpoint(lldb::SBBreakpoint &bp) {
```
There's a method `Breakpoint::CreateJsonObject` which basically does the same thing as this function. Can you reuse that one?

santhoshe447 replied:
Thanks,
I have not removed "CreateInstructionBreakpoint" function due to readability purpose, but just called the "CreateJsonObject" function from "CreateInstructionBreakpoint".

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Resolved follwoing review comments :
1. Removed usage of removed_ibp, removed deleted instruction breakpoints deirectly from g_dap.instruction_breakpoints.erase(prev_ibp.first)
2. Reused CreateJsonObject function in CreateInstructionBreakpoint.
3. Updated test cases with stop reason.

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Resolved code format issues.

santhoshe447 commented:
Hi Everyone,
Thanks for everyone's valuable feedback. Pls let me know if all are ok with the changes.

walter-erquinigo commented on lldb/tools/lldb-dap/JSONUtils.cpp:
```diff
@@ -893,7 +957,13 @@ llvm::json::Value CreateThreadStopped(lldb::SBThread &thread,
+      InstructionBreakpoint *inst_bp =
+          g_dap.GetInstructionBPFromStopReason(thread);
+      if (inst_bp) {
+        body.try_emplace("reason", "instruction breakpoint");
```
nice!

walter-erquinigo commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -4082,6 +4084,254 @@ void request__testGetTargetBreakpoints(const llvm::json::Object &request) {
+
+    // Iterate previouse active instruction breakpoint list.
+    for (auto &prev_ibp : g_dap.instruction_breakpoints) {
+      // Find previouse instruction breakpoint reference address in newly
```
```suggestion
      // Find previous instruction breakpoint reference address in newly
```

walter-erquinigo reviewed (approved):
This LGTM. I'd just want a rename from `instructionReference` to `addressReference` or `instructionAddressReference` for clarity.

santhoshe447 commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -996,4 +996,33 @@ void DAP::SetThreadFormat(llvm::StringRef format) {
 
+InstructionBreakpoint *
+DAP::GetInstructionBreakpoint(const lldb::break_id_t bp_id) {
+
```
I did not understand this.

walter-erquinigo replied:
The empty line :) Just to keep consistency with the rest of the codebase.

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Resolved review comments.

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Resolved review comments.

santhoshe447 commented on lldb/tools/lldb-dap/InstructionBreakpoint.h:
```diff
@@ -0,0 +1,36 @@
+// Instruction Breakpoint
+struct InstructionBreakpoint : public Breakpoint {
+
+  lldb::addr_t instructionReference;
```
renamed to "instructionAddressReference"  from instructionReference.

santhoshe447 committed:
[lldb-dap] Enabling instruction breakpoint support to lldb-dap.
Resolved review comments.

santhoshe447 committed:
Merge branch 'llvm:main' into main

DavidSpickett commented:
The rest of the tests are disabled on Windows, this one should be too.

One of these days I'll figure out what's wrong with them all.

santhoshe447 commented:
> The rest of the lldb-dap tests are disabled on Windows, this one should be too.
> 
> One of these days I'll figure out what's wrong with them all.

Thanks,
I have submitted the PR for this.
https://github.com/llvm/llvm-project/pull/106200
